### DISCUSSION TITLE
스트림 API 의 일회성

### DISCUSSION THREAD

**Discussion by corock** [Study-2-Modern-Java-In-Action/Study-2-Modern-Java-in-Action#46] [CHAPTER 4 스트림 소개 - Q&A]
스트림 API 는 왜 한 번만 소비할 수 있는가? 여러 번 소비하면 안 되는 이유가 궁금합니다.

**Comment by BvrPark** [comment]
* 컬렉션은 데이터 요소들을 메모리에 저장해서 사용하기 때문에 여러 번 소비를 할 수 있습니다.
* 하지만 스트림은 데이터의 시퀀스(데이터 요소들이 일련의 순서를 가지고 나열된 것)를 처리하기 위한 연속적인 요소들의 흐름이며, 
데이터 요소들을 저장하기 보단 단순히 읽어들입니다.
* 그래서 스트림API는 데이터를 저장하기보단 처리하는데 필요한 연산(중간 연산, 최종 연산)들을 제공하고 
한 번 처리한 데이터는 제거되기 때문에 다시 소비할 수가 없는 것 같습니다.
-> 여러 번 소비하면 안된다기 보단 설계 자체가 여러 번 소비할 수 없게 설계가 된 것 같습니다!

-> 비유하자면 세수할 때, 컬렉션은 물을 세면대에 받아서 사용하고 스트림은 흐르는 강물을 바로 사용하는 느낌이라 보면 되지 않을까요?

**요약**
컬렉션 : 데이터를 저장, 여러 번 소비가능
스트림 : 데이터를 읽음, 단순히 데이터를 읽는 것이므로 한 번 소비되면 다시 사용하는 것이 불가

**Reply by Irisation23** [reply]
먼저 저는 아래의 경우를 생각해 볼 수 있을 것 같습니다.

```
재사용 하려고 하는 Stream 코드가 비용적인 이점을 볼 수 있는가?
```

### 1. Stack Oveflow에 따르면 Stream과 Collection의 성능차이를 확인해 보면 Collection이 조금 더 빠른 결과를 도출합니다.
- [Stack Overflow 관련 자료 링크](https://stackoverflow.com/questions/22658322/java-8-performance-of-streams-vs-collections)

> 따라서, 단순 Stream의 사용이 코드 재사용 관점에서 비용적인 이점을 볼 수 있는가에 대한 관점에서는 회의적이라 생각합니다.

### 2. 병렬 스트림을 재사용 코드로 선택해 복잡한 고려사항을 모두 고려할 수 있는가? 

- [Oracle Blog 관련 자료 링크](https://blogs.oracle.com/javamagazine/post/java-parallel-streams-performance-benchmark) 
>  아직 학습되지 않은 내용이고 차후에 링크를 확인 해 보면 좋을 내용을 같이 담겠습니다.
